### PR TITLE
chore(generator): implement IFrame and IPage 

### LINKED ISSUE #1242: Use generated IFrame

### LINKED ISSUE #1245: Use generated IPage

### PR DESCRIPTION
**Pull request by kblok**
closes #1245 
closes #1242

### CODE REVIEW

File: src/PlaywrightSharp/Waiter.cs [hunk 1]
```diff
@@ -71,35 +71,12 @@ internal Task<T> WaitForEventAsync<T>(object eventSource, string e, Func<T, bool
...
 
-        internal Task WaitForEventAsync(object eventSource, string e)
+        internal Task<object> WaitForEventAsync(object eventSource, string e)
         {
-            var (task, dispose) = GetWaitForEventTask(eventSource, e);
+            var (task, dispose) = GetWaitForEventTask<object>(eventSource, e, null);
             return WaitForPromiseAsync(task, dispose);
         }
 
-        internal (Task Task, Action Dispose) GetWaitForEventTask(object eventSource, string eventName)
```
Comment (avodovnik) [inline src/PlaywrightSharp/Waiter.cs:80]: Yay. :D

File: src/PlaywrightSharp/Transport/Channels/RouteChannel.cs [hunk 2]
```diff
@@ -43,9 +43,9 @@ public Task ContinueAsync(string url, HttpMethod method, string postData, Dictio
                 args["method"] = method;
             }
 
-            if (!string.IsNullOrEmpty(postData))
+            if (postData != null)
             {
-                args["postData"] = Convert.ToBase64String(Encoding.UTF8.GetBytes(postData));
```
Comment (avodovnik) [inline src/PlaywrightSharp/Transport/Channels/RouteChannel.cs:48]: Just making a note of this - you already put this in the draft release notes, right?

File: src/PlaywrightSharp/PlaywrightSharp.csproj [hunk 3]
```diff
@@ -18,7 +18,7 @@
     <ReleaseVersion>0.0.0</ReleaseVersion>
     <BuildFromSource>True</BuildFromSource>
     <AnalysisMode>AllEnabledByDefault</AnalysisMode>
-    <NoWarn>1701;1702;CS0067</NoWarn>
+    <NoWarn>1701;1702;CS0067;1734;1574</NoWarn>
```
Comment (avodovnik) [inline src/PlaywrightSharp/PlaywrightSharp.csproj:21]: What are these?

File: src/PlaywrightSharp/PlaywrightSharp.csproj [hunk 4]
```diff
@@ -60,8 +60,6 @@
     </ContentWithTargetPath>
   </ItemGroup>
   <Target Name="EnsurePrerequisitsRan" BeforeTargets="InstallBrowsers">
-    <Error
-      Text="Playwright prerequisites are missing. Ensure you've ran `dotnet run -p ./src/tools/PlaywrightSharp.Tooling/PlaywrightSharp.Tooling.csproj -- download-drivers --basepath .`"
-      Condition="!Exists('$(MSBuildProjectDirectory)\Drivers\unix$(NativeFolder)')" />
+    <Error Text="Playwright prerequisites are missing. Ensure you've ran `dotnet run -p ./src/tools/PlaywrightSharp.Tooling/PlaywrightSharp.Tooling.csproj -- download-drivers --basepath .`" Condition="!Exists('$(MSBuildProjectDirectory)\Drivers\unix$(NativeFolder)')" />
```
Comment (avodovnik) [inline src/PlaywrightSharp/PlaywrightSharp.csproj:63]: Snuck that one in, did you? :-) (it's fine)

File: src/PlaywrightSharp/Contracts/IFrame.cs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        /// </para>
+        /// </summary>
+        /// <param name="expression">
+        /// JavaScript expression to be evaluated in the browser context. If it looks like a
+        /// function declaration, it is interpreted as a function. Otherwise, evaluated as an
+        /// expression.
+        /// </param>
+        /// <param name="arg">Optional argument to pass to <paramref name="expression"/>.</param>
+        /// <returns>A <see cref="Task"/> that will resolve when the evaluate function is executed by the browser.</returns>
+        public Task<JsonElement?> EvaluateAsync(string expression, object arg = default);
```
Comment (avodovnik) [inline src/PlaywrightSharp/Contracts/IFrame.cs:37]: This is purely to have a default value for `T` that returns a `JsonElement`? Do we *really* want to have this? It's a bunch of docs that needs to be maintained. At the very list, I'd reference the original function in the docs, using `<see>`

File: src/PlaywrightSharp/Contracts/IFrame.cs [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        /// for more details.
+        /// </param>
+        /// <param name="expression">
+        /// JavaScript expression to be evaluated in the browser context. If it looks like a
+        /// function declaration, it is interpreted as a function. Otherwise, evaluated as an
+        /// expression.
+        /// </param>
+        /// <param name="arg">Optional argument to pass to <paramref name="expression"/>.</param>
+        /// <returns>A <see cref="Task"/> that will resolve when the evaluate function is executed by the browser.</returns>
+        public Task EvalOnSelectorAllAsync(string selector, string expression, object arg);
```
Comment (avodovnik) [inline src/PlaywrightSharp/Contracts/IFrame.cs:63]: Same as above.

File: src/PlaywrightSharp/Contracts/IFrame.cs [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        Task<IReadOnlyCollection<string>> SelectOptionAsync(string selector, SelectOptionValue values, bool? noWaitAfter = default, float? timeout = default);
+
+        /// <inheritdoc cref="SelectOptionAsync(string, IEnumerable{SelectOptionValue}, bool?, float?)" />
+        Task<IReadOnlyCollection<string>> SelectOptionAsync(string selector, params string[] values);
+
+        /// <inheritdoc cref="SelectOptionAsync(string, IEnumerable{SelectOptionValue}, bool?, float?)" />
+        Task<IReadOnlyCollection<string>> SelectOptionAsync(string selector, params SelectOptionValue[] values);
+
+        /// <inheritdoc cref="SelectOptionAsync(string, IEnumerable{SelectOptionValue}, bool?, float?)" />
+        Task<IReadOnlyCollection<string>> SelectOptionAsync(string selector, params IElementHandle[] values);
```
Comment (avodovnik) [inline src/PlaywrightSharp/Contracts/IFrame.cs:299]: This feels like *too many* overloads. Are we sure we need them? Do we have any data backing up the need for so many? If it's purely convenience, I'd argue we should remove them, and wait for PRs/issues and/or move them to a `contrib` package.

**Review by avodovnik — COMMENTED** [review]
I'm reluctant in seeing so many overloads. I didn't comment on every single one of them, but the gist is there. Fundamentally, I ask myself - why do we need them? Is that an assumption that our users need them? Why are the other languages OK without them? Are we intentionally making the API surface area bigger by "assuming" what the users need? How painful would it be _for us_ to remove some (or most) of the overloads?

File: src/PlaywrightSharp/Transport/Channels/RouteChannel.cs (hunk 2 above)
Comment (kblok) [reply src/PlaywrightSharp/Transport/Channels/RouteChannel.cs:48]: Yes Sir!

File: src/PlaywrightSharp/PlaywrightSharp.csproj (hunk 3 above)
Comment (kblok) [reply src/PlaywrightSharp/PlaywrightSharp.csproj]: 1734 is stuff like this:
`XML comment on 'IDownload' has a paramref tag for 'acceptDownloads', but there is no parameter by that name (CS1734) (PlaywrightSharp)`
 `/// Browser context **must** be created with the <paramref name="acceptDownloads"/>`
I created [this issue](https://github.com/microsoft/playwright-sharp/issues/1304) to fix that.

I removed `1574`.

File: src/PlaywrightSharp/PlaywrightSharp.csproj (hunk 4 above)
Comment (kblok) [reply src/PlaywrightSharp/PlaywrightSharp.csproj]: IDE :/ 
restored.

File: src/PlaywrightSharp/Contracts/IFileChooser.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace PlaywrightSharp
+{
+    public partial interface IFileChooser
+    {
+        /// <inheritdoc cref="SetFilesAsync(System.Collections.Generic.IEnumerable{string}, bool?, float?)"/>
+        Task SetFilesAsync(string files, bool? noWaitAfter = default, float? timeout = default);
```
Comment (kblok) [inline src/PlaywrightSharp/Contracts/IFileChooser.cs:9]: With this we support the `[path]` overload defined here https://github.com/microsoft/playwright/blame/master/docs/src/api/params.md#L70

File: src/PlaywrightSharp/Contracts/IPage.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        IFrame FrameByUrl(string urlString);
+
+        /// <inheritdoc cref="FrameByUrl(string, Regex, Func{string, bool})"/>
+        IFrame FrameByUrl(Regex urlRegex);
+
+        /// <inheritdoc cref="FrameByUrl(string, Regex, Func{string, bool})"/>
+        IFrame FrameByUrl(Func<string, bool> urlFunc);
+
+        /// <inheritdoc cref="EmulateMediaAsync(Media, ColorScheme)"/>
+        Task EmulateMediaAsync();
```
Comment (avodovnik) [inline src/PlaywrightSharp/Contracts/IPage.cs:21]: I think this isn't needed.

**Review by avodovnik — APPROVED** [review]
Having thought more about this, my proposal is, let's keep most of the overloads you propose in _for the immediate future_, so we can potentially cut this week. We can fix the missing `inheritdocs` later. And we can take *some of the overloads* out when we get this and my incoming `IBrowserContext` PR done. What do you think?

**Review by avodovnik — CHANGES_REQUESTED** [review]
Let's take a look at the failing tests.

File: src/PlaywrightSharp.Tests/BrowserContextStorageStateTests.cs [hunk 10]
```diff
@@ -84,7 +84,7 @@ public void ShouldSetLocalStorage()
                 }
             });*/
             //var page1 = await context.NewPageAsync();
-            //await page1.RouteAsync("**/*", (route, _) =>
+            //await page1.RouteAsync("**/*", (route) =>
```
Comment (avodovnik) [inline src/PlaywrightSharp.Tests/BrowserContextStorageStateTests.cs:87]: can we just remove this commented code?

**Review by avodovnik — APPROVED** [review]
Maybe let's get rid of the commented code, but we can do that clean-up tasks later.

**Pull request merged**